```ruby
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

public class demo1 {

	
	public static void main(String[] args) {

	}
	
	
	
	
public static void read(){
	FileInputStream file1= null;
	
	try{

		
		File file=new File("D:\\abc.txt");
		file1= new FileInputStream(file);
		int length=0;
		 byte [] a=new byte[1024];
		while((length=file1.read(a))!=-1){
			
			System.out.println(new String(a,0,length));
			
			
		}
}catch(IOException e){
	
	throw  new RuntimeException(e);
	//组织后面的代码执行，并且通知调用者这里出问题
	
//但是return 不能提示信息。
	//System.exit(0);退出虚拟机更加不方便
	
	//所以用throw e;抛出异常，后面的代码就不会执行了,要声明异常，并且调用者要进行处理
	
	/*可以抛出 runtimeException() 运行时异常不用声明,可以包装一层，然后抛出，让调用者使用变得更加灵活。
	好似黄金里面包装了一个炸弹。方法不用声明，调用者可以处理也可以不处理*/
	
	
	//System.exit(0);//退出虚拟机
	
	
	}finally {
		//空管道不要关
		if (file1!=null){
		try{file1.close();}
		catch(IOException e){
			throw new RuntimeException(e);
			
		}
		}
	}
	
	
	
	
}

}

```
## 这是一个完整的处理异常的方法
```ruby
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class demo1 {

	public static void main(String[] args) {
	
		
		
   //复制文件
		copy();
		
		
		

	}
	
	
	public static void copy() {
		
		FileInputStream inputFile=null;
		FileOutputStream outfile=null;
		try{
			File file=new File("D:\\abc.txt");
			File file1=new File("D:\\a.txt");
		
		inputFile=new FileInputStream(file);
		 outfile=new FileOutputStream(file1);
		
		byte [] buf=new byte[1024];
		int length=0;
		while((length=inputFile.read(buf))!=-1){
			
			outfile.write(buf,0,length);
			
		}
		}catch(IOException e){
			System.out.println("拷贝出现问题");
			throw new RuntimeException(e);
			
		}finally{
			try{
				if(inputFile!=null){
			
			inputFile.close();
			System.out.println("关闭对象流成功");
				}
				
			}
			catch(IOException e){
				System.out.println("关闭输入流资源失败");
				throw new RuntimeException(e);
				
			}finally{
				
				try{
					if(outfile!=null){
				
				outfile.close();
				System.out.println("关闭输出流成功");
					}
					
				}
				catch(IOException e){
					System.out.println("关闭输出流资源失败");
					throw new RuntimeException(e);
			
				}
			}
		}
		
		
		
	}

}
